src/main/java/org/terasology/flyingIslands/FlyingIslandDensityProvider.java

```diff
@@ -32,28 +33,31 @@ public void process(GeneratingRegion region) {
         DensityFacet densityFacet = region.getRegionFacet(DensityFacet.class);
         SurfacesFacet surfacesFacet = region.getRegionFacet(SurfacesFacet.class);
 
-        for (Map.Entry<BaseVector3i, FlyingIsland> entry : flyingIslandFacet.getWorldEntries().entrySet()) {
+        for (Map.Entry<Vector3ic, FlyingIsland> entry : flyingIslandFacet.getWorldEntries().entrySet()) {
 
             Vector3i basePosition = new Vector3i(entry.getKey());
             FlyingIsland flyingIsland = entry.getValue();
 
             int extent = (int) flyingIsland.getOuterRadius();
-            int top = FlyingIsland.MAXHEIGHT;
+            int maxTop = FlyingIsland.MAX_DEPTH + FlyingIsland.MAX_HEIGHT;
 
             for (int i = -extent; i <= extent; i++) {
                 for (int k = -extent; k <= extent; k++) {
-                    Vector3i position = new Vector3i(i, top, k).add(basePosition);
+                    Vector3i position = new Vector3i(i, maxTop, k).add(basePosition);
 
-                    int height = flyingIsland.getHeightAndIsLava(position.x, position.z);
+                    int depth = flyingIsland.getDepth(position.x, position.z);
+                    int top = flyingIsland.getTop(position.x, position.z);

```

Comment:
You could make this section neater by having variables for the actual height of the top and bottom of the island in this column, rather than the height relative to the overall height of the island.